# Playlist Recommendation System

This notebook can be used to recommend songs when given a playlist of one or more songs. 

In [12]:
import numpy as np
import pandas as pd
from pandas.api.types import CategoricalDtype
from scipy.sparse import csr_matrix
from sklearn.metrics import pairwise_distances
from sklearn.metrics.pairwise import cosine_similarity

In [13]:
df = pd.read_json('pre_pivot.json')

In [14]:
df.shape

(515096, 13)

In [15]:
df.artist_and_track.nunique()

116157

In [16]:
df.artist_and_track.nunique()

116157

In [17]:
df.shape

(515096, 13)

In [18]:
df.added_by_id.nunique()

918

In [19]:
df.playlist_id.nunique()

6049

In [20]:
df.head()

added_by_id                album_id  \
0     brendan.ta  4uIDigk79DeZEYV6Z5Yf4s   
1     brendan.ta  4sFhah3DYcJlYeT47q3rhM   
10    brendan.ta  5cOhR878H8hC3UsxYq5Xyv   
100   brendan.ta  07hs3meNvGu5Fp46pnLQm7   
1000  brendan.ta  4PKH86wn7Gw4iel2WD564k   

                                             album_name  \
0                                        What Went Down   
1                       In The Silence (Deluxe Version)   
10    All My Friends (feat. Tinashe & Chance the Rap...   
100                                        Hungry Ghost   
1000                                     Do It Like You   

                                             artist_ids  \
0                              [6FQqZYVfTNQ1pCqfkwVFEa]   
1                              [7xUZ4069zcyBM4Bn10NQ1c]   
10    [2FwJwEswyIUAljqgjNSHgP, 0NIIxcxNHmOoyBx03SfTC...   
100                            [2N2EFVDEbp2JB8ulEUVIxp]   
1000                           [0aA1GTrIMutjIh4GlPPUVN]   

                                 artist_names            date_added  \
0                                     [Foals]  2016-03-26T00:48:16Z   
1                                    [Ásgeir]  2016-03-26T00:49:53Z   
10    [Snakehips, Tinashe, Chance the Rapper]  2016-04-29T05:34:47Z   
100                            [Violent Soho]  2017-02-16T03:09:55Z   
1000                        [Crooked Colours]  2019-02-04T05:22:30Z   

                 playlist_id                track_id  \
0     3pTPXB3vT93AOTSVozP54o  53L6A3I9vf7rgEZnMzx54E   
1     3pTPXB3vT93AOTSVozP54o  6VNo09sojPBi5mdckQkLbX   
10    3pTPXB3vT93AOTSVozP54o  6TaqooOXAEcijL6G1AWS2K   
100   3pTPXB3vT93AOTSVozP54o  14NnWDVM8nCssfHAqWZp1T   
1000  3LF6iEL6uCE9rSdLF4m1ga  5h1Gi50T9QxVfzvIF8vciC   

                                             track_name     user_id  rating  \
0                                  Mountain At My Gates  brendan.ta       1   
1                                        King and Cross  brendan.ta       1   
10    All My Friends (feat. Tinashe & Chance the Rap...  brendan.ta       1   
100                                   Covered In Chrome  brendan.ta       1   
1000                                     Do It Like You  brendan.ta       1   

                                 artists_join  \
0                                       Foals   
1                                      Ásgeir   
10    Snakehips___Tinashe___Chance the Rapper   
100                              Violent Soho   
1000                          Crooked Colours   

                                       artist_and_track  
0                        Foals|||||Mountain At My Gates  
1                             Ásgeir|||||King and Cross  
10    Snakehips___Tinashe___Chance the Rapper|||||Al...  
100                  Violent Soho|||||Covered In Chrome  
1000                 Crooked Colours|||||Do It Like You

| Users | Playlists   | Unique Songs   | Total Songs|
|------|------|------|------|
|   918  | 6049| 116157| 515096|

| Avg Playlists/User | Avg Songs/Playlist   |
|-----|------|
|   6.6  | 85.2|

In [21]:
def search_by_artist_and_track(df, artist, track):
    '''
    This function searches the database by artist and track. All tracks that contain substrings with an
    exact match of the artist and track strings are returned. Only the first match is printed.
    '''
    print(df[df.artist_and_track.str.contains(artist, regex=False, case=False) & df.artist_and_track.str.contains(track, regex=False, case=False)].artist_and_track.iloc[0])
    return df[df.artist_and_track.str.contains(artist, regex=False, case=False) & df.artist_and_track.str.contains(track, regex=False, case=False)]



In [22]:
def make_playlist(user_id, playlist_id, df, list_of_searches):
    '''
    Makes a playlist from a list of searches using the first songs returned by each search. Returns a dataframe
    representing the new playlist.
    '''
    new_df = pd.DataFrame()
    for artist,track in list_of_searches:
        curr = search_by_artist_and_track(df, artist, track).iloc[0]
        new_df = new_df.append(pd.DataFrame([[user_id, curr.album_id,curr.album_name,curr.artist_ids,
                   curr.artist_names,'NA',playlist_id,curr.track_id,
                   curr.track_name,user_id,1,curr.artists_join,curr.artist_and_track]], columns=df.columns))
    return new_df

In [23]:
def create_sparse(df):
    '''
    Creates a sparse matrix of the songs and playlists.
    '''
    playlist_id_c = CategoricalDtype(sorted(df.playlist_id.unique()), ordered=True)
    artist_and_track_c = CategoricalDtype(sorted(df.artist_and_track.unique()), ordered=True)

    row = df.playlist_id.astype(playlist_id_c).cat.codes
    col = df.artist_and_track.astype(artist_and_track_c).cat.codes
    sparse_matrix = csr_matrix((df['rating'], (row, col)), \
                               shape=(playlist_id_c.categories.size, artist_and_track_c.categories.size))
    return sparse_matrix, playlist_id_c

In [24]:
def create_sim_dict(playlist_id, df):
    '''
    Creates similarity dictionary from sparse matrix.
    '''
    sparse_matrix, playlist_id_c = create_sparse(df)
    similarities_playlists = cosine_similarity(sparse_matrix)
    playlists = np.array(playlist_id_c.categories)
    playlists_indices = list(playlists)
    sim_zip = zip(playlists_indices, similarities_playlists[playlists_indices.index(playlist_id)])
    sim_dict = {p:sim for p,sim in sim_zip}
    return sim_dict

In [25]:
def create_rec_df(playlist_id, df):
    '''
    Create recommendations with sum of similarities. This should be used over the mean of similarities if you
    want song popularity to influence song recommendations. Higher popularity means the song is more likely to
    be recommended. Returns recommendations, which are the songs listed in order of decreasing similarity sums.
    '''
    df_rec = df[df.playlist_id != playlist_id]
    sim_dict = create_sim_dict(playlist_id, df)
    df_sim = pd.DataFrame(zip(sim_dict.keys(), sim_dict.values()))
    df_sim.columns = ['playlist_id','sim']
    df_rec = df.merge(df_sim,how='left', on='playlist_id')
    grouped = df_rec.groupby(by = 'artist_and_track').sum()
    recommendations = grouped.sort_values(by = 'sim', ascending = False)
    return recommendations

In [26]:
def recommend_for_playlist(user_id, playlist_id, df, list_of_searches):
    '''
    Temporarily adds an input playlist into a dataframe in order to return similarities. Returns recommendations.
    Currently, it's likely that the input playlist songs are included in the recommendations. This is intentional
    in order to set up future work with the spotify api but may be changed in a later version.
    '''
    my_playlist = make_playlist(user_id, playlist_id, df, list_of_searches)
    df = df.append(my_playlist)
    recommendations = create_rec_df(playlist_id, df)
    return recommendations

In [27]:
def simplify(messy_rec, number_tracks):
    '''
    Creates a clean dataframe from recommendations, which is the return from the recommend_for_playlist functions.
    '''
    combined = []
    for row in messy_rec.head(number_tracks).itertuples():
        combined.append(row.Index.split('|||||'))
    clean_df = pd.DataFrame(np.array(combined))
    clean_df.columns = ['artist','track']
    return clean_df

In [28]:
def create_rec_df_mean(playlist_id, df):
    '''
    Create recommendations with mean of similarities. This should be used over the sum of similarities if you
    DO NOT want song popularity to influence song recommendations. This is a good option if you find you're
    getting the same recommendations often. Returns recommendations, which are the songs listed in order of
    decreasing similarity means.
    '''
    df_rec = df[df.playlist_id != playlist_id]
    sim_dict = create_sim_dict(playlist_id, df)
    df_sim = pd.DataFrame(zip(sim_dict.keys(), sim_dict.values()))
    df_sim.columns = ['playlist_id','sim']
    df_rec = df.merge(df_sim,how='left', on='playlist_id')
    grouped = df_rec.groupby(by = 'artist_and_track').mean()
    recommendations = grouped.sort_values(by = 'sim', ascending = False)
    return recommendations

In [29]:
def recommend_for_playlist_mean(user_id, playlist_id, df, list_of_searches):
    '''
    Temporarily adds an input playlist into a dataframe in order to return similarities. Returns recommendations.
    Currently, it's likely that the input playlist songs are included in the recommendations. This is intentional
    in order to set up future work with the spotify api but may be changed in a later version.
    '''
    my_playlist = make_playlist(user_id, playlist_id, df, list_of_searches)
    df = df.append(my_playlist)
    recommendations = create_rec_df_mean(playlist_id, df)
    return recommendations

In [33]:
simplify(recommend_for_playlist('justin', 'justins jams', df, [('croce','i got a name'),
                                                               ('steely dan','do it again'),
                                                              ('spinners','rubberband')]), 30)

Jim Croce|||||I Got A Name
Steely Dan|||||Do It Again
The Spinners|||||The Rubberband Man


artist                             track
0                   The Spinners                The Rubberband Man
1                     Steely Dan                       Do It Again
2                      Jim Croce                      I Got A Name
3       Electric Light Orchestra                      Mr. Blue Sky
4                  Looking Glass       Brandy (You're a Fine Girl)
5                  Rupert Holmes     Escape (The Pina Colada Song)
6   Creedence Clearwater Revival                     Fortunate Son
7            Yusuf / Cat Stevens                    Father And Son
8                         Boston               More Than a Feeling
9            The Doobie Brothers               Listen to the Music
10                 Fleetwood Mac         The Chain - 2004 Remaster
11      Electric Light Orchestra                      Livin' Thing
12           The Doobie Brothers                Long Train Runnin'
13               George Harrison                     My Sweet Lord
14      Electric Light Orchestra               Don't Bring Me Down
15            The Rolling Stones                     Gimme Shelter
16      Blue Swede___Björn Skifs               Hooked on a Feeling
17                    Elton John                       Tiny Dancer
18                    Steely Dan              Reelin' In The Years
19                        Eagles  Hotel California - 2013 Remaster
20                          10cc                   I'm Not In Love
21                 The Jackson 5                   I Want You Back
22                Lynyrd Skynyrd                         Free Bird
23                Stealers Wheel      Stuck In The Middle With You
24                        Kansas                  Dust in the Wind
25                   John Lennon              Imagine - Remastered
26                  Jimi Hendrix          All Along the Watchtower
27           Yusuf / Cat Stevens                        Wild World
28                  Elvin Bishop    Fooled Around And Fell In Love
29              Blue Öyster Cult           (Don't Fear) The Reaper

In [30]:
simplify(recommend_for_playlist_mean('justin', 'justins jams', df, [('croce','i got a name'),
                                                     ('steely dan','do it again')]), 20)

Jim Croce|||||I Got A Name
Steely Dan|||||Do It Again


artist  \
0                                 Jim Croce   
1                                Steely Dan   
2                       The Doobie Brothers   
3                             Dionne Farris   
4                          Titus Andronicus   
5                  Ambrosia___Tom Trefethen   
6                                   Chicago   
7                               Gary Wright   
8   Dickey Betts___The Allman Brothers Band   
9                            Linda Ronstadt   
10                        Steve Miller Band   
11                             Jay Ferguson   
12                               Elton John   
13                             Carl Carlton   
14                              Rod Stewart   
15                              Cheap Trick   
16                                Bob Seger   
17                               Billy Joel   
18                            The Guess Who   
19                       Jefferson Starship   

                                                track  
0                                        I Got A Name  
1                                         Do It Again  
2   Listen to the Music - Single Version; 2006 Rem...  
3                                              I Know  
4                                A More Perfect Union  
5                How Much I Feel - Remastered Version  
6                          Feelin' Stronger Every Day  
7                                        Dream Weaver  
8                                        Ramblin' Man  
9                                          Blue Bayou  
10                                  Fly Like an Eagle  
11                                     Thunder Island  
12                                Bennie And The Jets  
13                                   Everlasting Love  
14  Tonight's the Night (Gonna Be Alright) - 2009 ...  
15                                     California Man  
16                                    Beautiful Loser  
17                                           Pressure  
18                                     Share the Land  
19                                 Find Your Way Back

In [32]:
simplify(recommend_for_playlist_mean('Dan', 'Dans jams', df, [('led','stairway'),
                                                             ('paul','band on the run')]), 30)

Led Zeppelin|||||Stairway to Heaven - Remaster
Paul McCartney___Wings|||||Band On The Run
The Spinners|||||The Rubberband Man


artist  \
0                                        The Spinners   
1                              Paul McCartney___Wings   
2                                     The Revivalists   
3                                        Led Zeppelin   
4                                       Randal L Meek   
5                                         John Denver   
6                                       Natural Child   
7                                           Van Halen   
8                                              Boston   
9                                             Alabama   
10                                            Alabama   
11                                      The Crew Cuts   
12                                       Led Zeppelin   
13                                              Pilot   
14                                          Westerman   
15                                        Asaf Avidan   
16                                       Gabriel Rios   
17                                             Tamino   
18                                            Editors   
19                                              Sioen   
20                                                      
21                                            Reamonn   
22                                     Paul McCartney   
23                                        Nina Simone   
24                                              Pilot   
25           The West Coast Pop Art Experimental Band   
26  Perry Como___Mitchell Ayres & His Orchestra___...   
27                                       The Platters   
28                                   Gordon Lightfoot   
29                                         Whitesnake   

                                                track  
0                                  The Rubberband Man  
1                                     Band On The Run  
2                                            Stand Up  
3                       Stairway to Heaven - Remaster  
4                                Following Our Dreams  
5   Poems, Prayers and Promises - Live at the Univ...  
6                                           NSA Blues  
7                   (Oh) Pretty Woman - 2015 Remaster  
8                                           It's Easy  
9                                          Keepin' Up  
10                                 Richard Petty Fans  
11                                            Sh-Boom  
12                      Misty Mountain Hop - Remaster  
13                              Magic - 2003 Remaster  
14                                             Edison  
15                                 The Labyrinth Song  
16                                          Song no.7  
17                                              Tummy  
18                              No Sound but the Wind  
19                                            Cruisin  
20           Bastille - Locked Out of Heaven (STUBRU)  
21                                            Tonight  
22                            Live And Let Die - Live  
23                      Don't Let Me Be Misunderstood  
24                                              Magic  
25                                   I Won't Hurt You  
26  Magic Moments (with Mitchell Ayres and His Orc...  
27                                          My Prayer  
28                 The Wreck of the Edmund Fitzgerald  
29                        Here I Go Again - Radio Mix

In [24]:
recommend_for_playlist_mean('Dan', 'Dans jams', df, [('man','killers'),
                                                     ('bette','carnes'),
                                                     ('york groove','ace'),
                                                     ('follow me','muse'),
                                                     ('radio ga','queen'),
                                                     ('i want to break free','queen'),
                                                     ('xanadu','rush')])



The Killers|||||Human
Kim Carnes|||||Bette Davis Eyes
Ace Frehley|||||New York Groove
Muse|||||Follow Me
Queen|||||Radio Ga Ga - Remastered
Queen|||||I Want To Break Free - Remastered
Rush|||||Xanadu


rating       sim
artist_and_track                                                    
Queen|||||I Want To Break Free - Remastered              1  0.350852
Muse|||||Follow Me                                       1  0.223680
Rush|||||Xanadu                                          1  0.190906
Ace Frehley|||||New York Groove                          1  0.122969
T. Rex|||||Mambo Sun - Remastered                        1  0.070508
Kim Carnes|||||Bette Davis Eyes                          1  0.064655
KISS|||||God Of Thunder - Demo Version                   1  0.064628
KISS|||||Psycho Circus - Without Intro                   1  0.064628
KISS|||||Nothing Can Keep Me From You                    1  0.064628
KISS|||||A World Without Heroes                          1  0.064628
Queen|||||Radio Ga Ga - Remastered                       1  0.064241
KISS|||||Let's Put The X In Sex                          1  0.061867
KISS|||||Christine Sixteen                               1  0.061867
KISS|||||Reason To Live                                  1  0.061867
KISS|||||King Of The Night Time World - Live             1  0.055738
KISS|||||Into The Void                                   1  0.055738
Cornershop|||||Candyman                                  1  0.054534
KISS|||||Let Me Go, Rock 'N Roll                         1  0.052852
The Dead Weather|||||I Cut Like a Buffalo                1  0.052528
The Killers|||||Human                                    1  0.051606
Ryan Adams|||||Halloweenhead                             1  0.047246
The Specials|||||Enjoy Yourself (It's Later Tha...       1  0.047246
Gary Numan / Tubeway Army|||||Down in the Park           1  0.047246
Zero Boys|||||Civilization's Dying                       1  0.047246
Tones On Tail|||||Go! (Single Edit)                      1  0.047246
Parliament|||||Up For The Down Stroke                    1  0.047246
Young Fathers|||||GET UP                                 1  0.047246
Run–D.M.C.|||||Mary, Mary                                1  0.047246
Dead Boys|||||All This And More                          1  0.047246
KISS|||||Tears Are Falling                               1  0.046400
...                                                    ...       ...
Hogland___Vinil|||||Got You Babe                         1  0.000000
Hogland___Vinil___Johnning|||||Better Runaway            1  0.000000
Hogland___Tanjent|||||Missed Call                        1  0.000000
Hogland___Kevin Faye___Johnny Edlind|||||Unfoll...       1  0.000000
Hollie Col|||||I'm Not Calling                           1  0.000000
Hollie Cook|||||99                                       1  0.000000
Hollie Cook|||||Milk & Honey                             1  0.000000
Hollie Cook|||||Sugar Water                              1  0.000000
Holly Williams|||||Drinkin'                              1  0.000000
Holly Waxwing|||||Vibe                                   1  0.000000
Holly Throsby|||||Aeroplane                              1  0.000000
Holly McNarland|||||Numb                                 1  0.000000
Holly Johnson|||||Americanos                             1  0.000000
Holly Herndon|||||Eternal                                1  0.000000
Holly Golightly & The Brokeoffs|||||Devil Do             1  0.000000
Holly Drummond|||||Diving In - (Rameses B Remix)         1  0.000000
Holly Arrowsmith|||||Love Together                       1  0.000000
Hollow Light|||||Open Waters                             1  0.000000
Hollow Coves|||||We Will Run                             1  0.000000
Hollow Coves|||||These Memories                          1  0.000000
Hollow Coves|||||The Woods - Acoustic                    1  0.000000
Hollow Coves|||||The Woods                               1  0.000000
Hollow Coves|||||Ran Away                                1  0.000000
Hollow Coves|||||Home                                    1  0.000000
Hollow Coves|||||Coastline                               1  0.000000
Hollow Coves___Bass Fly___Laurent L|||||The Woo...

In [13]:
test_recs = recommend_for_playlist('JSON', 'jasons jams', df, [('future Islands', 'Dream'), ('Beach Fossils','Down')])



Future Islands|||||A Dream of You and Me
Beach Fossils|||||Down the Line


In [205]:
search_by_artist_and_track(df, 'Overnight','Maggie')

Maggie Rogers|||||Overnight


added_by_id                album_id               album_name  \
362930  if_u_like_playlists  5AHWNPo3gllDmixgAoFru4  Heard It In A Past Life   
500144     hipsterthemister  5AHWNPo3gllDmixgAoFru4  Heard It In A Past Life   
500242     hipsterthemister  5AHWNPo3gllDmixgAoFru4  Heard It In A Past Life   
719422              ram0nn7  5AHWNPo3gllDmixgAoFru4  Heard It In A Past Life   

                      artist_ids     artist_names            date_added  \
362930  [4NZvixzsSefsNiIqXn0NDe]  [Maggie Rogers]  2019-02-03T16:22:02Z   
500144  [4NZvixzsSefsNiIqXn0NDe]  [Maggie Rogers]  2019-01-19T16:30:21Z   
500242  [4NZvixzsSefsNiIqXn0NDe]  [Maggie Rogers]  2019-01-18T16:09:16Z   
719422  [4NZvixzsSefsNiIqXn0NDe]  [Maggie Rogers]  2019-01-24T18:12:58Z   

                   playlist_id                track_id track_name  \
362930  4smNayOKlCtwUAIPOfsKOK  4L9l5x0WuzNtyEBeqzvgPE  Overnight   
500144  2TwfFafDftf4WjpZTp00Sr  4L9l5x0WuzNtyEBeqzvgPE  Overnight   
500242  0dnGaIHv1Y85bQla9LvxXE  4L9l5x0WuzNtyEBeqzvgPE  Overnight   
719422  5QaYHdUWlOeRb4EVl2ipWv  4L9l5x0WuzNtyEBeqzvgPE  Overnight   

                    user_id  rating   artists_join  \
362930  if_u_like_playlists       1  Maggie Rogers   
500144     hipsterthemister       1  Maggie Rogers   
500242     hipsterthemister       1  Maggie Rogers   
719422      stoneygatesound       1  Maggie Rogers   

                   artist_and_track  
362930  Maggie Rogers|||||Overnight  
500144  Maggie Rogers|||||Overnight  
500242  Maggie Rogers|||||Overnight  
719422  Maggie Rogers|||||Overnight

In [185]:
simplify(jasons_recs, 20)

artist                            track
0          Beach Fossils                    Down the Line
1         Future Islands            A Dream of You and Me
2              Peach Pit                        Peach Pit
3            Tame Impala       The Less I Know The Better
4           Good Morning                       Warned You
5              boy pablo                        Everytime
6          Beach Fossils                  What a Pleasure
7            Tame Impala  Feels Like We Only Go Backwards
8            Mac DeMarco                 My Kind of Woman
9              The Drums                             Days
10           Summer Salt                      Sweet to Me
11           The Walters                    I Love You So
12           Banes World              You Say I'm in Love
13           Real Estate                      Had To Hear
14           No Vacation                          Yam Yam
15            Mazzy Star                    Fade Into You
16           Men I Trust                      Show Me How
17  Cigarettes After Sex                               K.
18           Beach House                       Space Song
19           Summer Salt                   Candy Wrappers

In [177]:
jasons_recs.head(20)

rating       sim
artist_and_track                                                 
Beach Fossils|||||Down the Line                      20  2.207815
Future Islands|||||A Dream of You and Me              4  1.205020
Peach Pit|||||Peach Pit                              20  0.619935
Tame Impala|||||The Less I Know The Better           99  0.452416
Good Morning|||||Warned You                          25  0.404688
boy pablo|||||Everytime                              25  0.375735
Beach Fossils|||||What a Pleasure                    13  0.369210
Tame Impala|||||Feels Like We Only Go Backwards      69  0.357343
Mac DeMarco|||||My Kind of Woman                     15  0.355247
The Drums|||||Days                                   14  0.328813
Summer Salt|||||Sweet to Me                          11  0.325796
The Walters|||||I Love You So                        16  0.322657
Banes World|||||You Say I'm in Love                  16  0.277858
Real Estate|||||Had To Hear                          14  0.273767
No Vacation|||||Yam Yam                              17  0.271614
Mazzy Star|||||Fade Into You                         74  0.271556
Men I Trust|||||Show Me How                          10  0.265700
Cigarettes After Sex|||||K.                          26  0.264084
Beach House|||||Space Song                           39  0.260280
Summer Salt|||||Candy Wrappers                       13  0.256296

In [152]:
coles_recs.head(20)

rating       sim  weighted?
artist_and_track                                                             
Masego___FKJ|||||Tadow                                54  0.479657   0.008883
Childish Gambino|||||Redbone                         127  0.443397   0.003491
Nick Hakim|||||Cuffed                                 19  0.346619   0.018243
Steve Lacy|||||Some                                   23  0.305193   0.013269
Matt Martians___Syd___Steve Lacy|||||Dent Jusay       10  0.300350   0.030035
Frank Ocean|||||Nikes                                 38  0.299221   0.007874
Masego|||||Navajo                                     27  0.283595   0.010504
BJ The Chicago Kid|||||Turnin' Me Up                  22  0.272489   0.012386
Sampa the Great|||||Blue Boss                         11  0.263032   0.023912
Jorja Smith|||||Teenage Fantasy                       15  0.260472   0.017365
Sir Froderick___DJ Harrison|||||Dumdums                8  0.258460   0.032308
Vulfpeck|||||Back Pocket                              23  0.249973   0.010868
SZA___Kendrick Lamar|||||Doves In The Wind            24  0.249955   0.010415
Nick Hakim|||||Roller Skates                           8  0.243969   0.030496
Thundercat|||||A Fan's Mail (Tron Song Suite II)      15  0.243498   0.016233
DJ Shadow|||||Midnight In A Perfect World             35  0.242472   0.006928
Nas|||||The World Is Yours                            44  0.241505   0.005489
KAYTRANADA___Syd|||||YOU'RE THE ONE                   28  0.240618   0.008593
SZA|||||Go Gina                                       13  0.240157   0.018474
Frank Ocean|||||Pink + White                          50  0.236721   0.004734

In [101]:
search_by_artist_and_track(df, 'Thundercat', 'Them Changes').iloc[0]

added_by_id                           patreeeek
album_id                 4anxEzh8xEKSPXKHGWiFue
album_name                                Drunk
artist_ids             [4frXpPxQQZwbCu3eTGnZEw]
artist_names                       [Thundercat]
date_added                 2018-06-16T20:27:17Z
playlist_id              0OLCESUYzhIboDmPv9XRGf
track_id                 5dcZUsgcMW57ju0A379O7c
track_name                         Them Changes
user_id                               patreeeek
rating                                        1
artists_join                         Thundercat
artist_and_track    Thundercat|||||Them Changes
Name: 123085, dtype: object

In [108]:
cole_df = make_playlist('test_5', 'doesnt matter', df, [('Thundercat', 'Them Changes'), ('Girlfriend Is Better','Talking Heads')])

In [109]:
cole_df

added_by_id                album_id                            album_name  \
0      test_5  4anxEzh8xEKSPXKHGWiFue                                 Drunk   
0      test_5  4sLCQxMRfn3gAHrBNZtbTH  Speaking In Tongues (Deluxe Version)   

                 artist_ids     artist_names date_added    playlist_id  \
0  [4frXpPxQQZwbCu3eTGnZEw]     [Thundercat]         NA  doesnt matter   
0  [2x9SpqnPi8rlE9pjHBwmSC]  [Talking Heads]         NA  doesnt matter   

                 track_id                                      track_name  \
0  5dcZUsgcMW57ju0A379O7c                                    Them Changes   
0  6VyR9z26oKw6BvlwSjqvoY  Girlfriend Is Better - 2005 Remastered Version   

  user_id  rating   artists_join  \
0  test_5       1     Thundercat   
0  test_5       1  Talking Heads   

                                    artist_and_track  
0                        Thundercat|||||Them Changes  
0  Talking Heads|||||Girlfriend Is Better - 2005 ...

In [95]:
search_by_artist_and_track(df, 'Girlfriend Is Better', 'Talking Heads')

added_by_id                album_id  \
142614                  patreeeek  4sLCQxMRfn3gAHrBNZtbTH   
161383                  patreeeek  78MM8HrabEGPLVWaJkM2t1   
163370                  patreeeek  4sLCQxMRfn3gAHrBNZtbTH   
423606               cannondale12  17K4P45WAJU1GeCqAIC8OE   
543316  21n2ku53disfrt53dgeg3vbki  78MM8HrabEGPLVWaJkM2t1   

                                  album_name                artist_ids  \
142614  Speaking In Tongues (Deluxe Version)  [2x9SpqnPi8rlE9pjHBwmSC]   
161383                   Speaking In Tongues  [2x9SpqnPi8rlE9pjHBwmSC]   
163370  Speaking In Tongues (Deluxe Version)  [2x9SpqnPi8rlE9pjHBwmSC]   
423606  Speaking In Tongues (Deluxe Version)  [2x9SpqnPi8rlE9pjHBwmSC]   
543316                   Speaking In Tongues  [2x9SpqnPi8rlE9pjHBwmSC]   

           artist_names            date_added             playlist_id  \
142614  [Talking Heads]  2019-01-20T01:58:35Z  2HJy3TNqrRXr6DMvXFGCxN   
161383  [Talking Heads]  2013-03-09T15:41:44Z  0KRN0GjJaYn4uc2daHTWBx   
163370  [Talking Heads]  2019-04-02T17:59:20Z  5CzjSMJf8z9pp2ek0xCGmM   
423606  [Talking Heads]  2014-03-14T00:37:16Z  428zCMVCQGMmbCCYawqsBV   
543316  [Talking Heads]  2017-02-17T22:43:15Z  58fnmHO3IapTcllIDF4P4p   

                      track_id  \
142614  6VyR9z26oKw6BvlwSjqvoY   
161383  37kkHqpkzTEGOJYHRb6JZG   
163370  6VyR9z26oKw6BvlwSjqvoY   
423606  2UYY46dgNshqDfDF20RBYM   
543316  37kkHqpkzTEGOJYHRb6JZG   

                                            track_name  \
142614  Girlfriend Is Better - 2005 Remastered Version   
161383                            Girlfriend Is Better   
163370  Girlfriend Is Better - 2005 Remastered Version   
423606  Girlfriend Is Better - 2005 Remastered Version   
543316                            Girlfriend Is Better   

                          user_id  rating   artists_join  \
142614                  patreeeek       1  Talking Heads   
161383                  patreeeek       1  Talking Heads   
163370                  patreeeek       1  Talking Heads   
423606               cannondale12       1  Talking Heads   
543316  21n2ku53disfrt53dgeg3vbki       1  Talking Heads   

                                         artist_and_track  
142614  Talking Heads|||||Girlfriend Is Better - 2005 ...  
161383             Talking Heads|||||Girlfriend Is Better  
163370  Talking Heads|||||Girlfriend Is Better - 2005 ...  
423606  Talking Heads|||||Girlfriend Is Better - 2005 ...  
543316             Talking Heads|||||Girlfriend Is Better

In [92]:
search_by_artist_and_track(df, 'Thundercat', 'Them Changes')

added_by_id                album_id  \
123085                  patreeeek  4anxEzh8xEKSPXKHGWiFue   
150551                  patreeeek  4anxEzh8xEKSPXKHGWiFue   
216399                 1126708296  1atZ1TiIPZZec0oiyY1mBl   
326714                 1214486915  4anxEzh8xEKSPXKHGWiFue   
345045               hilltopchill  7vgnnPWflv8g0vQYOberLn   
346332                   odrey918  7vHBQDqwzB7uDvoE5bncMM   
393411                tmacgordon5  4anxEzh8xEKSPXKHGWiFue   
411012                   devotron  4anxEzh8xEKSPXKHGWiFue   
416680                 1259677891  4anxEzh8xEKSPXKHGWiFue   
543763  21n2ku53disfrt53dgeg3vbki  1atZ1TiIPZZec0oiyY1mBl   
566867  212tnazcrrn7pprlaztfbdfzq  7vHBQDqwzB7uDvoE5bncMM   
579810  212tnazcrrn7pprlaztfbdfzq  7vHBQDqwzB7uDvoE5bncMM   
612219                 1231174271  04HVQ56JChI6cgMVdx1s0w   
612275                 1231174271  1atZ1TiIPZZec0oiyY1mBl   
612836                 1231174271  2MBRta0pAW0J0Wshh4cGNN   
618135                 1231174271  2MBRta0pAW0J0Wshh4cGNN   
618157                 1231174271  1atZ1TiIPZZec0oiyY1mBl   
620774                 1231174271  4anxEzh8xEKSPXKHGWiFue   
632809                 jmgsimpson  4anxEzh8xEKSPXKHGWiFue   
634548                  tom_misch  1atZ1TiIPZZec0oiyY1mBl   
635773               ellielynwood  1atZ1TiIPZZec0oiyY1mBl   
63846                 zerochances  1atZ1TiIPZZec0oiyY1mBl   
641994                 1279593689  4anxEzh8xEKSPXKHGWiFue   
64293                 zerochances  1atZ1TiIPZZec0oiyY1mBl   
64908                 zerochances  1atZ1TiIPZZec0oiyY1mBl   
64963                 zerochances  1atZ1TiIPZZec0oiyY1mBl   
65136                 zerochances  1atZ1TiIPZZec0oiyY1mBl   
666937                 1255019854  4anxEzh8xEKSPXKHGWiFue   
667457                 1255019854  4anxEzh8xEKSPXKHGWiFue   
667608                 1255019854  4anxEzh8xEKSPXKHGWiFue   
667984                 1255019854  7vgnnPWflv8g0vQYOberLn   
668669                 1255019854  4anxEzh8xEKSPXKHGWiFue   
669631                 1255019854  4anxEzh8xEKSPXKHGWiFue   
671831                 1255019854  7vgnnPWflv8g0vQYOberLn   
678230                  jaysnbrwn  2MBRta0pAW0J0Wshh4cGNN   
679850                willzinkann  7vHBQDqwzB7uDvoE5bncMM   
680103                willzinkann  4anxEzh8xEKSPXKHGWiFue   
682224                willzinkann  4anxEzh8xEKSPXKHGWiFue   
687928                willzinkann  7vHBQDqwzB7uDvoE5bncMM   
688225                willzinkann  4anxEzh8xEKSPXKHGWiFue   
690811                 genius8086  1atZ1TiIPZZec0oiyY1mBl   
693365  83vwkvb6l83s1exev977gilgr  4anxEzh8xEKSPXKHGWiFue   
693483  83vwkvb6l83s1exev977gilgr  4anxEzh8xEKSPXKHGWiFue   
721773            carson.genger55  7vHBQDqwzB7uDvoE5bncMM   
744621             jeeveswilliams  1atZ1TiIPZZec0oiyY1mBl   

                                album_name  \
123085                               Drunk   
150551                               Drunk   
216399  The Beyond / Where the Giants Roam   
326714                               Drunk   
345045  The Beyond / Where the Giants Roam   
346332                               Drunk   
393411                               Drunk   
411012                               Drunk   
416680                               Drunk   
543763  The Beyond / Where the Giants Roam   
566867                               Drunk   
579810                               Drunk   
612219                               Drunk   
612275  The Beyond / Where the Giants Roam   
612836  The Beyond / Where the Giants Roam   
618135  The Beyond / Where the Giants Roam   
618157  The Beyond / Where the Giants Roam   
620774                               Drunk   
632809                               Drunk   
634548  The Beyond / Where the Giants Roam   
635773  The Beyond / Where the Giants Roam   
63846   The Beyond / Where the Giants Roam   
641994                               Drunk   
64293   The Beyond / Where the Giants Roam   
64908   The Beyond / Where the Giants Roam   
64963   The Beyond 

In [69]:
df2 = pd.DataFrame([['test_added2', '5CZR6ljD0x9fTiS4mh9wMp','Only By The Night','[2qk9voo8llSGYcZ6xrBzKx]','[Kings of Leon]','2017-08-23T18:14:04Z','test_playlist2','0ntQJM78wzOLVeCUAW7Y45','Sex on Fire','test_user2',1,'Kings of Leon','Kings of Leon|||||Sex on Fire']], columns=df.columns)
df2

added_by_id                album_id         album_name  \
0  test_added2  5CZR6ljD0x9fTiS4mh9wMp  Only By The Night   

                 artist_ids     artist_names            date_added  \
0  [2qk9voo8llSGYcZ6xrBzKx]  [Kings of Leon]  2017-08-23T18:14:04Z   

      playlist_id                track_id   track_name     user_id  rating  \
0  test_playlist2  0ntQJM78wzOLVeCUAW7Y45  Sex on Fire  test_user2       1   

    artists_join               artist_and_track  
0  Kings of Leon  Kings of Leon|||||Sex on Fire

In [70]:
df2.dtypes

added_by_id         object
album_id            object
album_name          object
artist_ids          object
artist_names        object
date_added          object
playlist_id         object
track_id            object
track_name          object
user_id             object
rating               int64
artists_join        object
artist_and_track    object
dtype: object

In [71]:
df = df.append(df2)

In [73]:
sparse_matrix

<6051x116157 sparse matrix of type '<class 'numpy.int64'>'
	with 489351 stored elements in Compressed Sparse Row format>

In [74]:
similarities_playlists = cosine_similarity(sparse_matrix)

In [75]:
playlists = np.array(playlist_id_c.categories)

In [76]:
playlists_indices = list(playlists)

In [77]:
playlists_indices

['002MBUoZVJntSIsgW6EpFA',
 '0061U3t8hRFhIBw42to0Yw',
 '008wDZ2VZxGTRYLYS82mya',
 '00AP7VyekxNKmWiMj9xisM',
 '00CsN1rCRTN8tKcPKcHCjQ',
 '00CzKdtJgJelMbJJvs2rJN',
 '00LP85DvUnR9CHeqAFv7VL',
 '00SrfsM0Wh4ZfZSrA9OQS3',
 '00UWUyGDpyN7bhsoUZVqdo',
 '00Vv9ktcShVGl695Nqtqjn',
 '00WPgq2AQvZ6vaXnaUddy0',
 '00b9NrNWGW7kB3mlk6lTUd',
 '00nBF6hbRoRLUKp2WvRrtF',
 '00rjDqOCrMtgV7wcxScgHM',
 '00wHFIu00pH4CJWPV4vJwA',
 '01AAC8Hn9BqioBD4AITbYN',
 '01AyypY3Vg6BZ7Fvbpi8AO',
 '01Bj96ebCNSskOmkh8taPP',
 '01GwqHIWoJmhLKfRhu5uvq',
 '01Ivk66sjv7fxG28mSEC3B',
 '01JIbtNjFwWjuddNL0Tq71',
 '01U7nKpzTRK1Hp2V03IMWu',
 '01UbuTuMY0TvMz42KXoy3y',
 '01VpWWi8xzzS6nJ64J5hQA',
 '01WaTuwNZnqwbBkRk2UeG9',
 '01aCMPN9vISPWsMqQtol8S',
 '01gjluhMJeFT0WcvR1937w',
 '01gyv6qrcXgycqwnbhMHO2',
 '01oOxpTnnokNxu5AitOL41',
 '01qHdqpm6S7zv5N3vChwye',
 '01rNo73R3xjRpa3G0VbKVH',
 '01tvqPxen74Nz7TRz7kpSd',
 '01wyjmhDW4Fru5cb0TQ7Dp',
 '023J3zoqmmqR0cUp7hyAcw',
 '02NUIaAZADXg8qYnwCezev',
 '02Tt4tSbfwN8EYTMc3ZzgM',
 '02Yypb6ck9z09yWm41yU5l',
 

In [78]:
playlists_indices.index('test_playlist2')

6050

In [79]:
similarities_playlists[6050]

array([0., 0., 0., ..., 0., 0., 1.])

In [17]:
# np.save('playlist_similarities',similarities_playlists)

In [20]:
test_dict = create_sim_dict('test_playlist')

In [52]:
recs = create_rec_df('test_playlist', df)

In [55]:
recs

rating       sim
artist_and_track                                                    
Kendrick Lamar___Jay Rock|||||Money Trees               43  4.764937
Kendrick Lamar|||||Bitch, Don’t Kill My Vibe            54  1.814687
Kendrick Lamar|||||HUMBLE.                              95  1.379705
Kendrick Lamar|||||DNA.                                 76  1.341237
Kendrick Lamar|||||ELEMENT.                             39  1.219696
Kendrick Lamar|||||Backseat Freestyle                   37  1.202267
Kendrick Lamar___Zacari|||||LOVE. FEAT. ZACARI.         74  1.158540
Kendrick Lamar|||||Swimming Pools (Drank) - Ext...      32  1.139217
Kendrick Lamar___MC Eiht|||||m.A.A.d city               52  1.103397
Kendrick Lamar|||||King Kunta                           72  1.092789
Kendrick Lamar___Drake|||||Poetic Justice               42  1.063346
Kendrick Lamar|||||A.D.H.D                              40  1.054424
Kendrick Lamar|||||Alright                              50  1.006350
A$AP Rocky___Drake___2 Chainz___Kendrick Lamar|...      44  0.985999
21 Savage|||||a lot                                     67  0.948533
J. Cole|||||No Role Modelz                              60  0.928979
Childish Gambino|||||Redbone                           127  0.901054
Kendrick Lamar___Rihanna|||||LOYALTY. FEAT. RIH...      40  0.900650
A$AP Rocky___Skepta|||||Praise The Lord (Da Shi...      86  0.868972
Kendrick Lamar___Bilal___Anna Wise___Thundercat...      39  0.828860
J. Cole|||||MIDDLE CHILD                               100  0.812161
Kendrick Lamar|||||FEAR.                                27  0.808350
Travis Scott|||||goosebumps                             81  0.805040
Dr. Dre___Snoop Dogg|||||The Next Episode               69  0.781475
Kendrick Lamar___JAY Z|||||Bitch, Don’t Kill My...      13  0.780621
Jay Rock___Kendrick Lamar___Future___James Blak...      44  0.762487
J. Cole|||||Neighbors                                   45  0.762273
Kendrick Lamar___Dr. Dre|||||The Recipe - Bonus...      16  0.741858
Kendrick Lamar|||||DUCKWORTH.                           25  0.724572
Travis Scott|||||STARGAZING                             54  0.699807
...                                                    ...       ...
Grouper|||||Heavy Water/I'd Rather Be Sleeping           5  0.000000
Grouper|||||Headache                                     3  0.000000
Grouper|||||Alien Observer                               2  0.000000
Grouplove|||||Let Me In                                  5  0.000000
Grouplove|||||Let Me In - Soundtrack Version             3  0.000000
Grouplove|||||Love Will Save Your Soul                   2  0.000000
Grouplove|||||Ways To Go                                29  0.000000
Grover Washington, Jr.|||||Just the Two of Us            2  0.000000
Grover Washington, Jr.|||||Inner City Blues (Ma...       2  0.000000
Grover Washington, Jr.|||||In the Name of Love           3  0.000000
Grover Washington, Jr.|||||Hydra                         3  0.000000
Grover Washington, Jr.|||||Black Frost                   3  0.000000
Grover Washington, Jr.___Bill Withers|||||Just ...       8  0.000000
Grover Washington, Jr.___Bill Withers|||||Just ...       3  0.000000
Grover Washington, Jr.___Bill Withers|||||Just ...       2  0.000000
Grover Anderson|||||Standing Water                       2  0.000000
Grouplove|||||Welcome To Your Life                      20  0.000000
Grouplove|||||Traumatized                                4  0.000000
Grouplove|||||Lovely Cup                                 8  0.000000
Grouplove|||||Torso                                      2  0.000000
Grouplove|||||Tongue Tied                               59  0.000000
Grouplove|||||Tell Me a Story                            3  0.000000
Grouplove|||||Spun                                       8  0.000000
Grouplove|||||Shark Attack                              10  0.000000
Grouplove|||||Remember That Night                        7  0.000000
Grouplove|||||Raspberry                           

In [32]:
df.head()

added_by_id                album_id  \
0     brendan.ta  4uIDigk79DeZEYV6Z5Yf4s   
1     brendan.ta  4sFhah3DYcJlYeT47q3rhM   
10    brendan.ta  5cOhR878H8hC3UsxYq5Xyv   
100   brendan.ta  07hs3meNvGu5Fp46pnLQm7   
1000  brendan.ta  4PKH86wn7Gw4iel2WD564k   

                                             album_name  \
0                                        What Went Down   
1                       In The Silence (Deluxe Version)   
10    All My Friends (feat. Tinashe & Chance the Rap...   
100                                        Hungry Ghost   
1000                                     Do It Like You   

                                             artist_ids  \
0                              [6FQqZYVfTNQ1pCqfkwVFEa]   
1                              [7xUZ4069zcyBM4Bn10NQ1c]   
10    [2FwJwEswyIUAljqgjNSHgP, 0NIIxcxNHmOoyBx03SfTC...   
100                            [2N2EFVDEbp2JB8ulEUVIxp]   
1000                           [0aA1GTrIMutjIh4GlPPUVN]   

                                 artist_names            date_added  \
0                                     [Foals]  2016-03-26T00:48:16Z   
1                                    [Ásgeir]  2016-03-26T00:49:53Z   
10    [Snakehips, Tinashe, Chance the Rapper]  2016-04-29T05:34:47Z   
100                            [Violent Soho]  2017-02-16T03:09:55Z   
1000                        [Crooked Colours]  2019-02-04T05:22:30Z   

                 playlist_id                track_id  \
0     3pTPXB3vT93AOTSVozP54o  53L6A3I9vf7rgEZnMzx54E   
1     3pTPXB3vT93AOTSVozP54o  6VNo09sojPBi5mdckQkLbX   
10    3pTPXB3vT93AOTSVozP54o  6TaqooOXAEcijL6G1AWS2K   
100   3pTPXB3vT93AOTSVozP54o  14NnWDVM8nCssfHAqWZp1T   
1000  3LF6iEL6uCE9rSdLF4m1ga  5h1Gi50T9QxVfzvIF8vciC   

                                             track_name     user_id  rating  \
0                                  Mountain At My Gates  brendan.ta       1   
1                                        King and Cross  brendan.ta       1   
10    All My Friends (feat. Tinashe & Chance the Rap...  brendan.ta       1   
100                                   Covered In Chrome  brendan.ta       1   
1000                                     Do It Like You  brendan.ta       1   

                                 artists_join  \
0                                       Foals   
1                                      Ásgeir   
10    Snakehips___Tinashe___Chance the Rapper   
100                              Violent Soho   
1000                          Crooked Colours   

                                       artist_and_track  
0                        Foals|||||Mountain At My Gates  
1                             Ásgeir|||||King and Cross  
10    Snakehips___Tinashe___Chance the Rapper|||||Al...  
100                  Violent Soho|||||Covered In Chrome  
1000                 Crooked Colours|||||Do It Like You

In [86]:
recs = create_rec_df('test_playlist2', df)

In [87]:
recs

rating       sim
artist_and_track                                                    
Kings of Leon|||||Sex on Fire                           57  5.506253
The Killers|||||Mr. Brightside                         107  1.757159
Kings of Leon|||||Use Somebody                          49  1.657102
The White Stripes|||||Seven Nation Army                 88  1.534124
Franz Ferdinand|||||Take Me Out                         90  1.457202
The Killers|||||Somebody Told Me                        53  1.209054
The Strokes|||||Last Nite                               65  1.195187
Arctic Monkeys|||||Do I Wanna Know?                     89  1.070059
Red Hot Chili Peppers|||||Californication               82  1.047806
Red Hot Chili Peppers|||||Snow (Hey Oh)                 65  1.007030
The Black Keys|||||Lonely Boy                           60  1.002935
Two Door Cinema Club|||||What You Know                  63  0.991522
Spin Doctors|||||Two Princes                            59  0.985916
Gorillaz|||||Feel Good Inc                              71  0.974046
Red Hot Chili Peppers|||||Can't Stop                    61  0.966273
Kaiser Chiefs|||||Ruby                                  52  0.948276
Foo Fighters|||||The Pretender                          51  0.941780
Nirvana|||||Come As You Are                             93  0.930239
Foster The People|||||Pumped Up Kicks                   88  0.914244
Muse|||||Uprising                                       45  0.901593
MGMT|||||Kids                                           72  0.899647
Radiohead|||||Creep                                     75  0.886566
Jet|||||Are You Gonna Be My Girl                        47  0.886508
Coldplay|||||Viva La Vida                               68  0.885397
Linkin Park|||||Numb                                    56  0.877683
Modest Mouse|||||Float On                               55  0.854405
R.E.M.|||||Losing My Religion                           78  0.851749
Train|||||Drops of Jupiter                              40  0.845500
Smash Mouth|||||All Star                                82  0.829301
Nirvana|||||Smells Like Teen Spirit                    107  0.829163
...                                                    ...       ...
I.B.M.|||||Kill Bill                                     2  0.000000
I-Octane|||||We Love Di Vibes                            2  0.000000
I-Octane|||||We Got Life                                 2  0.000000
I-Octane|||||Trend Setta                                 2  0.000000
IAN SWEET|||||#23                                        3  0.000000
ICEHOUSE|||||Crazy                                       3  0.000000
IDLES|||||1049 Gotho                                     2  0.000000
ICEHOUSE|||||Electric Blue                               2  0.000000
IDK|||||Windows Up                                       3  0.000000
IDK|||||Trigger Happy                                    3  0.000000
IDK|||||No Words                                         2  0.000000
IDK|||||No Shoes On The Rug, Leave Them At The ...       2  0.000000
IDK|||||Blame My Friends (The Gang)                      2  0.000000
IDK|||||Black Sheep, White Dove                          2  0.000000
IDK|||||Birds & The Bees                                 2  0.000000
IDK___Yung Gleesh|||||Baby Scale (with Yung Gle...       2  0.000000
IDK___Maxo Kream|||||Moral (feat. Maxo Kream)            4  0.000000
IDK___Joey Bada$$___Russ|||||Lil Arrogant (feat...       4  0.000000
IDK___Denzel Curry|||||ONCE UPON A TIME (FREEST...       6  0.000000
IDK___Denzel Curry|||||No Wave                           2  0.000000
IDER|||||Mirror                                          3  0.000000
IDER|||||King Ruby                                       2  0.000000
ID46|||||Bass On The Floor                               2  0.000000
ID16 ft. Vilde Lie|||||Buena Vista 2013                  6  0.000000
ICYTWAT|||||Z4 Music                                     3  0.000000
ICYTWAT|||||Shame                                 